test/extended/authorization/rolebinding_restrictions.go

```diff
@@ -0,0 +1,226 @@
+
+				// Subject not bound, not permitted by any RBR, rolebinding should fail
+				roleBindingCreate(oc, false, true, ns, user2, "rb3")
+
+				// Subject not bound, not permitted by any RBR, RBAC rolebinding should fail
+				roleBindingCreate(oc, true, true, ns, user2, "rb3")
+
+				// Create a rolebinding that also contains system:non-existing users should succeed
+				_, err = oc.AdminAuthorizationClient().AuthorizationV1().RoleBindingRestrictions(ns).Create(generateRBRnonExist(ns, user3))
+				o.Expect(err).NotTo(o.HaveOccurred())
+
+				roleBindingCreate(oc, false, false, ns, user3, "rb4")
+
+				_, err = oc.AdminAuthorizationClient().AuthorizationV1().RoleBindings(ns).Create(generateRolebindingNonExisting(ns, "rb5"))
+				o.Expect(err).NotTo(o.HaveOccurred())
+			})
+		})
+	})
+})
+
+func generateAllowUserRolebindingRestriction(ns, user string) *authorizationv1.RoleBindingRestriction {
+	return &authorizationv1.RoleBindingRestriction{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("match-users-%s", user),
+			Namespace: ns,
+		},
+		Spec: authorizationv1.RoleBindingRestrictionSpec{
+			UserRestriction: &authorizationv1.UserRestriction{
+				Users: []string{user},
+			},
+		},
+	}
+}
+
+func generateRBRnonExist(ns, user string) *authorizationv1.RoleBindingRestriction {
+	return &authorizationv1.RoleBindingRestriction{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("match-users-%s-and-non-existing", user),
+			Namespace: ns,
+		},
+		Spec: authorizationv1.RoleBindingRestrictionSpec{
+			UserRestriction: &authorizationv1.UserRestriction{
+				Users: []string{user, "system:non-existing"},
+			},
+		},
+	}
+}
+
+func generateRolebinding(ns, user, rb string) *authorizationv1.RoleBinding {
+	return &authorizationv1.RoleBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: ns,
+			Name:      rb,
+		},
+		Subjects: []corev1.ObjectReference{
+			{
+				Kind:      authorizationapi.UserKind,
+				Namespace: ns,
+				Name:      user,
+			},
+		},
+		RoleRef: corev1.ObjectReference{Name: "role", Namespace: ns},
+	}
+}
+
+func generateRbacUserRolebinding(ns, user, rb string) *rbacv1.RoleBinding {
+	return &rbacv1.RoleBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: ns,
+			Name:      rb,
+		},
+		Subjects: []rbacv1.Subject{
+			{
+				Kind:      rbacv1.UserKind,
+				Namespace: ns,
+				Name:      user,
+			},
+		},
+		RoleRef: rbacv1.RoleRef{Kind: "Role", Name: "role"},
+	}
+}
+
+func generateRolebindingNonExisting(ns, rb string) *authorizationv1.RoleBinding {
+	return &authorizationv1.RoleBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: ns,
+			Name:      rb,
+		},
+		Subjects: []corev1.ObjectReference{
+			{
+				Kind:      authorizationapi.UserKind,
+				Namespace: ns,
+				Name:      "system:non-existing",
+			},
+		},
+		RoleRef: corev1.ObjectReference{Name: "role", Namespace: ns},
+	}
+}
+
+func roleBindingCreate(oc *exutil.CLI, useRBAC, shouldErr bool, ns, user, rb string) {
```

Comment:
I think this is a bit easier to read:

```go
func roleBindingCreate(oc *exutil.CLI, useRBAC, shouldErr bool, ns, user, rb string) {
	var rbrErr error
	switch {
	case useRBAC:
		_, rbrErr = oc.AdminKubeClient().RbacV1().RoleBindings(ns).Create(generateRbacUserRolebinding(ns, user, rb))
	default:
		_, rbrErr = oc.AdminAuthorizationClient().AuthorizationV1().RoleBindings(ns).Create(generateRolebinding(ns, user, rb))
	}
	if !shouldErr {
		o.Expect(rbrErr).NotTo(o.HaveOccurred())
		return
	}
	o.Expect(rbrErr).NotTo(BeNil())
	o.Expect(kerrors.IsForbidden(rbrErr)).To(o.BeTrue())
	expectedErrorString := fmt.Sprintf("rolebindings to User \"%s\" are not allowed in project", user)
	o.Expect(rbrErr.Error()).Should(o.ContainSubstring(expectedErrorString))
}
```

Reply:
yes, it is, thanks.  I didn't realize you reviewed this so I've pushed some cleanup to the e2e test, please will you take a look again at this file? thanks :slightly_smiling_face:

Reply:
also, I've made that change